lib/galaxy/tools/cache.py

```diff
@@ -79,31 +84,33 @@ def get_tool_by_id(self, tool_id):
         return self.get_tool(self._tool_paths_by_id.get(tool_id))
 
     def expire_tool(self, tool_id):
-        if tool_id in self._tool_paths_by_id:
-            config_filename = self._tool_paths_by_id[tool_id]
-            del self._hash_by_tool_paths[config_filename]
-            del self._tool_paths_by_id[tool_id]
-            del self._tools_by_path[config_filename]
-            del self._mod_time_by_path[config_filename]
-            if tool_id in self._new_tool_ids:
-                self._new_tool_ids.remove(tool_id)
+        with self._lock:
+            if tool_id in self._tool_paths_by_id:
+                config_filename = self._tool_paths_by_id[tool_id]
+                del self._hash_by_tool_paths[config_filename]
+                del self._tool_paths_by_id[tool_id]
+                del self._tools_by_path[config_filename]
+                del self._mod_time_by_path[config_filename]
+                if tool_id in self._new_tool_ids:
+                    self._new_tool_ids.remove(tool_id)
 
     def cache_tool(self, config_filename, tool):
-        tool_hash = md5_hash_file(config_filename)
-        if tool_hash is None:
-            return
-        tool_id = str(tool.id)
-        self._hash_by_tool_paths[config_filename] = tool_hash
-        self._mod_time_by_path[config_filename] = os.path.getmtime(config_filename)
-        self._tool_paths_by_id[tool_id] = config_filename
-        self._tools_by_path[config_filename] = tool
-        self._new_tool_ids.add(tool_id)
-        for macro_path in tool._macro_paths:
-            self._mod_time_by_path[macro_path] = os.path.getmtime(macro_path)
-            if tool_id not in self._macro_paths_by_id:
-                self._macro_paths_by_id[tool_id] = {macro_path}
-            else:
-                self._macro_paths_by_id[tool_id].add(macro_path)
+        with self._lock:
+            tool_hash = md5_hash_file(config_filename)
+            if tool_hash is None:
+                return
+            tool_id = str(tool.id)
+            self._hash_by_tool_paths[config_filename] = tool_hash
+            self._mod_time_by_path[config_filename] = os.path.getmtime(config_filename)
+            self._tool_paths_by_id[tool_id] = config_filename
+            self._tools_by_path[config_filename] = tool
+            self._new_tool_ids.add(tool_id)
+            for macro_path in tool._macro_paths:
+                self._mod_time_by_path[macro_path] = os.path.getmtime(macro_path)
+                if tool_id not in self._macro_paths_by_id:
+                    self._macro_paths_by_id[tool_id] = {macro_path}
+                else:
+                    self._macro_paths_by_id[tool_id].add(macro_path)
 
     def reset_status(self):
         """
```

Comment:
Not relevant for the unit test failure, but probably also this method should use the lock.

Reply:
Yep, was just looking at that too.